File: federator/src/protocol.ts
/**
 * Protocol definitions for federator-node communication
 */

// Node types in the hybrid architecture
export type NodeType = 'cache' | 'storage';

// Node status
export type NodeStatus = 'healthy' | 'degraded' | 'offline';

// Cache statistics sent with heartbeats
export interface CacheStats {
  objects_count: number;
  total_size_bytes: number;
  hit_rate: number;
  requests_total: number;
}

// Object metadata
export interface ObjectMeta {
  bucket: string;
  key: string;
  etag: string;
  size: number;
  owner_node?: string;    // Storage node that owns this object (if any)
  cache_nodes: string[];  // Cache nodes that have a copy
}

// Node information
export interface NodeInfo {
  id: string;
  node_type: NodeType;
  buckets: string[];
  endpoint?: string;      // For storage nodes - HTTP endpoint for proxying
  last_heartbeat: number;
  status: NodeStatus;
}

// Bucket information
export interface BucketInfo {
  name: string;
  node_ids: string[];     // Nodes that have registered this bucket
  object_count: number;
}

// =============================================================================
// Node → Federator Messages
// =============================================================================

export interface AuthMessage {
  type: 'auth';
  jwt: string;
}

export interface RegisterMessage {
  type: 'register';
  node_id: string;
  node_type: NodeType;
  buckets: string[];
  endpoint?: string;      // Required for storage nodes
}

export interface HeartbeatMessage {
  type: 'heartbeat';
  node_id: string;
  stats: CacheStats;
}

export interface ObjectPutMessage {
  type: 'object_put';
  bucket: string;
  key: string;
  etag: string;
  size: number;
  is_owner: boolean;      // true if this is a storage node claiming ownership
}

export interface ObjectDeleteMessage {
  type: 'object_delete';
  bucket: string;
  key: string;
}

export interface BucketAddMessage {
  type: 'bucket_add';
  bucket: string;
}

export interface BucketRemoveMessage {
  type: 'bucket_remove';
  bucket: string;
}

export interface LocateObjectMessage {
  type: 'locate_object';
  bucket: string;
  key: string;
}

export type NodeMessage =
  | AuthMessage
  | RegisterMessage
  | HeartbeatMessage
  | ObjectPutMessage
  | ObjectDeleteMessage
  | BucketAddMessage
  | BucketRemoveMessage
  | LocateObjectMessage;

// =============================================================================
// Federator → Node Messages
// =============================================================================

export interface AuthOkMessage {
  type: 'auth_ok';
}

export interface AuthFailedMessage {
  type: 'auth_failed';
  reason: string;
}

export interface RegisteredMessage {
  type: 'registered';
  cluster_id: string;
}

export interface InvalidateMessage {
  type: 'invalidate';
  bucket: string;
  key: string;
  source_node: string;
}

export interface BucketSyncMessage {
  type: 'bucket_sync';
  bucket: string;
  objects: ObjectMeta[];
}

export interface NodeStatusMessage {
  type: 'node_status';
  nodes: NodeInfo[];
}

export interface LocateResponseMessage {
  type: 'locate_response';
  bucket: string;
  key: string;
  owner_node?: string;
  endpoint?: string;
}

export interface ErrorMessage {
  type: 'error';
  code: string;
  message: string;
}

export type FederatorMessage =
  | AuthOkMessage
  | AuthFailedMessage
  | RegisteredMessage
  | InvalidateMessage
  | BucketSyncMessage
  | NodeStatusMessage
  | LocateResponseMessage
  | ErrorMessage;

// =============================================================================
// Validation Functions
// =============================================================================

export function isNodeMessage(msg: unknown): msg is NodeMessage {
  if (typeof msg !== 'object' || msg === null) return false;
  const m = msg as Record<string, unknown>;
  if (typeof m.type !== 'string') return false;

  switch (m.type) {
    case 'auth':
      return typeof m.jwt === 'string';
    case 'register':
      return (
        typeof m.node_id === 'string' &&
        (m.node_type === 'cache' || m.node_type === 'storage') &&
        Array.isArray(m.buckets)
      );
    case 'heartbeat':
      return typeof m.node_id === 'string' && typeof m.stats === 'object';
    case 'object_put':
      return (
        typeof m.bucket === 'string' &&
        typeof m.key === 'string' &&
        typeof m.etag === 'string' &&
        typeof m.size === 'number' &&
        typeof m.is_owner === 'boolean'
      );
    case 'object_delete':
      return typeof m.bucket === 'string' && typeof m.key === 'string';
    case 'bucket_add':
    case 'bucket_remove':
      return typeof m.bucket === 'string';
    case 'locate_object':
      return typeof m.bucket === 'string' && typeof m.key === 'string';
    default:
      return false;
  }
}

// JWT payload structure
export interface NodeJwtPayload {
  node_id: string;
  cluster_id: string;
  iat: number;
  exp: number;
}

File: src/federator/protocol.rs
//! Protocol definitions for federator-node communication
//!
//! Message types match the TypeScript federator implementation.

use serde::{Deserialize, Serialize};

// Re-export NodeType from config
pub use crate::config::NodeType;

/// Node status
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum NodeStatus {
    Healthy,
    Degraded,
    Offline,
}

/// Cache statistics sent with heartbeats
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CacheStats {
    pub objects_count: u64,
    pub total_size_bytes: u64,
    pub hit_rate: f64,
    pub requests_total: u64,
}

/// Object metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObjectMeta {
    pub bucket: String,
    pub key: String,
    pub etag: String,
    pub size: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub owner_node: Option<String>,
    pub cache_nodes: Vec<String>,
}

/// Node information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NodeInfo {
    pub id: String,
    pub node_type: NodeType,
    pub buckets: Vec<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub endpoint: Option<String>,
    pub last_heartbeat: u64,
    pub status: NodeStatus,
}

// =============================================================================
// Node → Federator Messages
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum NodeMessage {
    Auth {
        jwt: String,
    },
    Register {
        node_id: String,
        node_type: NodeType,
        buckets: Vec<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        endpoint: Option<String>,
    },
    Heartbeat {
        node_id: String,
        stats: CacheStats,
    },
    ObjectPut {
        bucket: String,
        key: String,
        etag: String,
        size: u64,
        is_owner: bool,
    },
    ObjectDelete {
        bucket: String,
        key: String,
    },
    BucketAdd {
        bucket: String,
    },
    BucketRemove {
        bucket: String,
    },
    LocateObject {
        bucket: String,
        key: String,
    },
}

// =============================================================================
// Federator → Node Messages
// =============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum FederatorMessage {
    AuthOk,
    AuthFailed {
        reason: String,
    },
    Registered {
        cluster_id: String,
    },
    Invalidate {
        bucket: String,
        key: String,
        source_node: String,
    },
    BucketSync {
        bucket: String,
        objects: Vec<ObjectMeta>,
    },
    NodeStatus {
        nodes: Vec<NodeInfo>,
    },
    LocateResponse {
        bucket: String,
        key: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        owner_node: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        endpoint: Option<String>,
    },
    Error {
        code: String,
        message: String,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_node_message_serialization() {
        let msg = NodeMessage::Auth {
            jwt: "test-token".to_string(),
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("\"type\":\"auth\""));
        assert!(json.contains("\"jwt\":\"test-token\""));
    }

    #[test]
    fn test_register_message_serialization() {
        let msg = NodeMessage::Register {
            node_id: "node-1".to_string(),
            node_type: NodeType::Cache,
            buckets: vec!["bucket-1".to_string()],
            endpoint: None,
        };
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("\"type\":\"register\""));
        assert!(json.contains("\"node_type\":\"cache\""));
    }

    #[test]
    fn test_federator_message_deserialization() {
        let json = r#"{"type":"auth_ok"}"#;
        let msg: FederatorMessage = serde_json::from_str(json).unwrap();
        assert!(matches!(msg, FederatorMessage::AuthOk));
    }

    #[test]
    fn test_invalidate_message_deserialization() {
        let json = r#"{"type":"invalidate","bucket":"test","key":"obj1","source_node":"node-2"}"#;
        let msg: FederatorMessage = serde_json::from_str(json).unwrap();
        if let FederatorMessage::Invalidate {
            bucket,
            key,
            source_node,
        } = msg
        {
            assert_eq!(bucket, "test");
            assert_eq!(key, "obj1");
            assert_eq!(source_node, "node-2");
        } else {
            panic!("Expected Invalidate message");
        }
    }
}
